### PR TITLE
Refactored DockerFile to improve efficiency

### LINKED ISSUE #2596: Admin: Extremely Poor Docker Resource Utilization Efficiency
**Describe the bug**

We run a demonstration instance of Talawa-Admin on a GoDaddy VPS server running Ubuntu. It has the following resources:

1. 1 core
1. 4 GB of RAM
1. 40 GB of disk

Other information:

1. The demo instance is intended to create an evaluation environment for new GtiHub contributors and users alike as they decide to use Talawa. The DB of the demo instance gets reset every day.
2. Talawa Admin runs natively on this VPS server with acceptable performance with one user. The load average is approximately 1, which is the target value for a system with only 1 core.
3. When Talawa Admin runs on the server using docker. The load average reaches 130, the swap process is the top CPU resource user. The system is so overloaded that only one ssh session at a time is achievable.

The purpose of this issue is to find ways to tune all Talawa-Admin Dockerfile and app configurations to lower its CPU and RAM utilization by at least 75% 

1. With the current Docker performance very few developers or end users will want to try Talawa themselves.
2. This has been a recurring issue with Talawa Admin. The poor performance threatens the success of our current MongoDB based MVP.

**To Reproduce**

1. Run Talawa-Admin on a system
2. See excessive resource utilization

**Expected behavior**

1. Acceptable usage information such that it can run easily on a mid-range laptop without impacting its performance

**Actual behavior**

1. Poor performance

**Screenshots**

image

**Additional details**

**Potential internship candidates**

Please read this if you are planning to apply for a Palisadoes Foundation internship 
- https://github.com/PalisadoesFoundation/talawa/issues/359

### PR DESCRIPTION
**Pull request by adithyanotfound**
**What kind of change does this PR introduce?**

Refactoring

**Issue Number:**

Fixes #2596 

**Did you add tests for your changes?**

Not required

**Snapshots/Videos:**

CPU and RAM utilization

Screenshot 2024-12-05 at 9 18 20 PM

Docker image size

Screenshot 2024-12-05 at 11 07 07 PM
<br></br>

**Summary**

1. Reduce Docker image size from 1.75gb to 66mb
2. Improved performance of app using docker 
3. Used nginx and multi-stage build for scaling the app and reduce image size

**Does this PR introduce a breaking change?**

Yes

**Other information**

TO-DO
1. Add ssl certificate to nginx.conf
2. Configure nginx reverse proxy properly to handle api requests

**Have you read the [contributing guide](https://github.com/PalisadoesFoundation/talawa-admin/blob/master/CONTRIBUTING.md)?**

Yes

## Summary by CodeRabbit

- **New Features**
	- Introduced a new service configuration in Docker for streamlined application deployment.
	- Added a new Nginx configuration to enhance server performance and security.

- **Bug Fixes**
	- Improved health check for the application to ensure better reliability.

- **Chores**
	- Updated dependencies in the project to maintain compatibility and improve functionality.
	- Added entries to `.eslintignore` to exclude specific files from linting.

### CODE REVIEW

File: package.json [hunk 1]
```diff
@@ -69,7 +68,9 @@
     "vite-plugin-environment": "^1.1.3",
     "vite-plugin-node-polyfills": "^0.22.0",
     "vite-tsconfig-paths": "^5.1.3",
-    "web-vitals": "^4.2.4"
+    "web-vitals": "^4.2.4",
+    "vite-plugin-svgr": "^4.2.0",
+    "cross-env": "^7.0.3"
```
[the remark this answers is not in the document]
Comment (adithyanotfound) [reply package.json]: made the required changes

File: docker-compose.yml [hunk 2]
```diff
@@ -0,0 +1,11 @@
...
+  app:
+    build:
+      context: .
+    # Ideally these would be dynamically setup by the end user similar to .env
+    ports:
+      - '4321:80'
+    environment:
+      - REACT_APP_TALAWA_URL=http://localhost:4000/graphql/
+      - REACT_APP_BACKEND_WEBSOCKET_URL=ws://localhost:4000/graphql/
+      - PORT=4321
```
[the remark this answers is not in the document]
Comment (adithyanotfound) [reply docker-compose.yml:11]: Added the suggestions

File: Dockerfile [hunk 3]
```diff
@@ -4,12 +1,22 @@
...
 COPY package*.json ./
-
 RUN npm install
 
 COPY . .
 
-RUN npm run build
+ENV NODE_ENV=production
+
+RUN yarn build
```
[the remark this answers is not in the document]
Comment (adithyanotfound) [reply Dockerfile:12]: Used npm instead of yarn

File: nginx.conf [hunk 4]
```diff
@@ -0,0 +1,27 @@
...
+        # CORS should be made strict before deployment (currently allows access from any origin)
+        add_header Access-Control-Allow-Origin *;
+        add_header Access-Control-Allow-Methods "GET, POST, OPTIONS";
+        add_header Access-Control-Allow-Headers "Content-Type, Authorization";
+        proxy_http_version 1.1;
+        proxy_set_header Upgrade $http_upgrade;
+        proxy_set_header Connection "upgrade";
+        proxy_set_header Host $host;
+        proxy_cache_bypass $http_upgrade;
+    }
```
[the remark this answers is not in the document]
Comment (adithyanotfound) [reply nginx.conf:23]: Added the changes

File: Dockerfile [hunk 5]
```diff
@@ -1,15 +1,22 @@
...
 
-EXPOSE 4321
+ENV NODE_ENV=production
 
-CMD ["npm", "run", "serve"]
\ No newline at end of file
+COPY nginx.conf /etc/nginx/conf.d/default.conf
+COPY --from=builder /talawa-admin/build /usr/share/nginx/html
+EXPOSE 80
+CMD ["nginx", "-g", "daemon off;"]
```
[the remark this answers is not in the document]
Comment (adithyanotfound) [reply Dockerfile]: will add later

**Review by palisadoes — CHANGES_REQUESTED** [review]
Please move the `nginx.conf` to a `config/docker/setup` directory so that it is very clearly a part of the setup / installation process.  That way it is less likely to be deleted or edited in error

File: config/docker/setup/nginx.conf [hunk 6]
```diff
@@ -0,0 +1,48 @@
+server {
+    listen 80;
+    server_name ${NGINX_SERVER_NAME};
+    
+    # TODO: Add SSL configuration
+    # listen 443 ssl;
+    # ssl_certificate /etc/nginx/ssl/cert.pem;
+    # ssl_certificate_key /etc/nginx/ssl/key.pem;
+
```
[the remark this answers is not in the document]
Comment (adithyanotfound) [reply config/docker/setup/nginx.conf:9]: need it only for prod not now

File: config/docker/setup/nginx.conf [hunk 7]
```diff
@@ -0,0 +1,48 @@
...
+    root /usr/share/nginx/html;
+    index index.html;
+
+    add_header X-Frame-Options "DENY";
+    add_header X-Content-Type-Options "nosniff";
+    add_header X-XSS-Protection "1; mode=block";
+    add_header Referrer-Policy "strict-origin-when-cross-origin";
+    # add_header Content-Security-Policy "default-src 'self'; script-src 'self'; style-src 'self'; connect-src 'self' https://api.com;";
+
```
[the remark this answers is not in the document]
Comment (adithyanotfound) [reply config/docker/setup/nginx.conf:18]: not needed for now

**Pull request merged**